File: src/main/java/org/example/kalkulationsprogramm/controller/LieferantReklamationController.java
package org.example.kalkulationsprogramm.controller;

import lombok.RequiredArgsConstructor;
import org.example.kalkulationsprogramm.domain.*;
import org.example.kalkulationsprogramm.dto.CreateReklamationRequest;
import org.example.kalkulationsprogramm.dto.LieferantReklamationDto;
import org.example.kalkulationsprogramm.repository.*;
import org.example.kalkulationsprogramm.controller.LieferantenController.LieferantBildDto;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.util.StringUtils;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import java.util.Objects;
import java.util.UUID;
import java.util.stream.Collectors;

@RestController
@RequestMapping("/api/reklamationen")
@RequiredArgsConstructor
public class LieferantReklamationController {

    private final LieferantReklamationRepository reklamationRepository;
    private final LieferantenRepository lieferantenRepository;
    private final LieferantDokumentRepository dokumentRepository;
    private final MitarbeiterRepository mitarbeiterRepository;
    private final LieferantBildRepository bildRepository;

    @GetMapping("/lieferant/{lieferantId}")
    public ResponseEntity<List<LieferantReklamationDto>> getByLieferant(@PathVariable Long lieferantId) {
        if (!lieferantenRepository.existsById(lieferantId)) {
            return ResponseEntity.notFound().build();
        }
        var list = reklamationRepository.findByLieferantIdOrderByStatusAscErstelltAmDesc(lieferantId);
        return ResponseEntity.ok(list.stream().map(this::mapToDto).collect(Collectors.toList()));
    }

    @PatchMapping("/{id}/status")
    @Transactional
    public ResponseEntity<Void> updateStatus(@PathVariable Long id, @RequestParam ReklamationStatus status) {
        var reklamation = reklamationRepository.findById(id).orElse(null);
        if (reklamation == null) {
            return ResponseEntity.notFound().build();
        }
        reklamation.setStatus(status);
        reklamationRepository.save(reklamation);
        return ResponseEntity.ok().build();
    }

    @DeleteMapping("/{id}")
    @Transactional
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        if (!reklamationRepository.existsById(id)) {
            return ResponseEntity.notFound().build();
        }
        reklamationRepository.deleteById(id);
        return ResponseEntity.noContent().build();
    }

    @PostMapping("/lieferant/{lieferantId}")
    @Transactional
    public ResponseEntity<LieferantReklamationDto> create(
            @PathVariable Long lieferantId,
            @RequestBody CreateReklamationRequest request,
            @RequestParam(required = false) String token) {

        var lieferant = lieferantenRepository.findById(lieferantId).orElse(null);
        if (lieferant == null) {
            return ResponseEntity.notFound().build();
        }

        var mitarbeiter = mitarbeiterByToken(token);

        LieferantReklamation reklamation = new LieferantReklamation();
        reklamation.setLieferant(lieferant);
        reklamation.setErstelltVon(mitarbeiter);
        reklamation.setBeschreibung(request.getBeschreibung());
        reklamation.setStatus(request.getStatus() != null ? request.getStatus() : ReklamationStatus.OFFEN);

        if (request.getLieferscheinId() != null) {
            var lieferschein = dokumentRepository.findById(request.getLieferscheinId()).orElse(null);
            if (lieferschein != null && lieferschein.getLieferant().getId().equals(lieferantId)) {
                reklamation.setLieferschein(lieferschein);
            }
        }

        reklamation = reklamationRepository.save(reklamation);
        return ResponseEntity.status(HttpStatus.CREATED).body(mapToDto(reklamation));
    }

    @PostMapping(value = "/{id}/bilder", consumes = MediaType.MULTIPART_FORM_DATA_VALUE)
    @Transactional
    public ResponseEntity<LieferantBildDto> uploadBild(
            @PathVariable Long id,
            @RequestPart("datei") MultipartFile datei,
            @RequestParam(required = false) String token) {

        var reklamation = reklamationRepository.findById(id).orElse(null);
        if (reklamation == null) {
            return ResponseEntity.notFound().build();
        }

        var mitarbeiter = mitarbeiterByToken(token);

        try {
            String originalFilename = StringUtils.cleanPath(Objects.requireNonNull(datei.getOriginalFilename()));
            String storedFilename = UUID.randomUUID() + "_" + originalFilename;

            // Reuse the existing supplier images folder structure
            Path lieferantDir = Paths.get("uploads", "lieferanten",
                    reklamation.getLieferant().getId().toString(), "bilder");
            Files.createDirectories(lieferantDir);
            Path targetPath = lieferantDir.resolve(storedFilename);
            datei.transferTo(targetPath);

            LieferantBild bild = new LieferantBild();
            bild.setLieferant(reklamation.getLieferant());
            bild.setReklamation(reklamation); // Link to reclamation
            bild.setOriginalDateiname(originalFilename);
            bild.setGespeicherterDateiname(storedFilename);
            bild.setErstelltAm(java.time.LocalDateTime.now());
            bild.setHochgeladenVon(mitarbeiter);

            bild = bildRepository.save(bild);

            LieferantBildDto dto = new LieferantBildDto();
            dto.id = bild.getId();
            dto.originalDateiname = bild.getOriginalDateiname();
            dto.url = "/api/lieferanten/bilder/file/" + bild.getGespeicherterDateiname();
            dto.erstelltAm = bild.getErstelltAm();
            if (mitarbeiter != null) {
                dto.mitarbeiterVorname = mitarbeiter.getVorname();
                dto.mitarbeiterNachname = mitarbeiter.getNachname();
            }
            return ResponseEntity.status(HttpStatus.CREATED).body(dto);

        } catch (IOException e) {
            return ResponseEntity.internalServerError().build();
        }
    }

    @GetMapping("/lieferscheine/search")
    public ResponseEntity<List<LieferscheinSearchDto>> searchLieferscheine(
            @RequestParam Long lieferantId,
            @RequestParam String query) {

        var results = dokumentRepository.searchLieferscheine(lieferantId, query);
        var dtos = results.stream().map(doc -> {
            LieferscheinSearchDto dto = new LieferscheinSearchDto();
            dto.setId(doc.getId());
            dto.setOriginalDateiname(doc.getEffektiverDateiname());
            if (doc.getGeschaeftsdaten() != null) {
                dto.setDokumentNummer(doc.getGeschaeftsdaten().getDokumentNummer());
                dto.setDatum(doc.getGeschaeftsdaten().getDokumentDatum());
            } else {
                dto.setDatum(doc.getUploadDatum().toLocalDate());
            }
            return dto;
        }).collect(Collectors.toList());

        return ResponseEntity.ok(dtos);
    }

    private LieferantReklamationDto mapToDto(LieferantReklamation entity) {
        LieferantReklamationDto dto = new LieferantReklamationDto();
        dto.setId(entity.getId());
        dto.setLieferantId(entity.getLieferant().getId());
        dto.setLieferantName(entity.getLieferant().getLieferantenname());

        if (entity.getLieferschein() != null) {
            dto.setLieferscheinId(entity.getLieferschein().getId());
            dto.setLieferscheinDateiname(entity.getLieferschein().getEffektiverDateiname());
            if (entity.getLieferschein().getGeschaeftsdaten() != null) {
                dto.setLieferscheinNummer(entity.getLieferschein().getGeschaeftsdaten().getDokumentNummer());
            }
        }

        if (entity.getErstelltVon() != null) {
            dto.setErstellerName(entity.getErstelltVon().getVorname() + " " + entity.getErstelltVon().getNachname());
        }

        dto.setErstelltAm(entity.getErstelltAm());
        dto.setBeschreibung(entity.getBeschreibung());
        dto.setStatus(entity.getStatus());

        dto.setBilder(entity.getBilder().stream().map(b -> {
            LieferantBildDto img = new LieferantBildDto();
            img.id = b.getId();
            img.originalDateiname = b.getOriginalDateiname();
            img.url = "/api/lieferanten/bilder/file/" + b.getGespeicherterDateiname();
            img.beschreibung = b.getBeschreibung();
            img.erstelltAm = b.getErstelltAm();
            if (b.getHochgeladenVon() != null) {
                img.mitarbeiterVorname = b.getHochgeladenVon().getVorname();
                img.mitarbeiterNachname = b.getHochgeladenVon().getNachname();
            }
            return img;
        }).collect(Collectors.toList()));

        return dto;
    }

    private Mitarbeiter mitarbeiterByToken(String token) {
        if (!StringUtils.hasText(token)) {
            return null;
        }
        return mitarbeiterRepository.findByLoginToken(token).orElse(null);
    }

    @lombok.Data
    public static class LieferscheinSearchDto {
        private Long id;
        private String dokumentNummer;
        private String originalDateiname;
        private java.time.LocalDate datum;
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/controller/LieferantReklamationController.kt
package org.example.kalkulationsprogramm.controller

import org.example.kalkulationsprogramm.controller.LieferantenController.LieferantBildDto
import org.example.kalkulationsprogramm.domain.LieferantBild
import org.example.kalkulationsprogramm.domain.LieferantReklamation
import org.example.kalkulationsprogramm.domain.Mitarbeiter
import org.example.kalkulationsprogramm.domain.ReklamationStatus
import org.example.kalkulationsprogramm.dto.CreateReklamationRequest
import org.example.kalkulationsprogramm.dto.LieferantReklamationDto
import org.example.kalkulationsprogramm.repository.LieferantBildRepository
import org.example.kalkulationsprogramm.repository.LieferantDokumentRepository
import org.example.kalkulationsprogramm.repository.LieferantReklamationRepository
import org.example.kalkulationsprogramm.repository.LieferantenRepository
import org.example.kalkulationsprogramm.repository.MitarbeiterRepository
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.transaction.annotation.Transactional
import org.springframework.util.StringUtils
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RequestPart
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.multipart.MultipartFile
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.Objects
import java.util.UUID

@RestController
@RequestMapping("/api/reklamationen")
class LieferantReklamationController(
    private val reklamationRepository: LieferantReklamationRepository,
    private val lieferantenRepository: LieferantenRepository,
    private val dokumentRepository: LieferantDokumentRepository,
    private val mitarbeiterRepository: MitarbeiterRepository,
    private val bildRepository: LieferantBildRepository,
) {
    @GetMapping("/lieferant/{lieferantId}")
    fun getByLieferant(@PathVariable lieferantId: Long): ResponseEntity<List<LieferantReklamationDto>> {
        if (!lieferantenRepository.existsById(lieferantId)) {
            return ResponseEntity.notFound().build()
        }
        val list = reklamationRepository.findByLieferantIdOrderByStatusAscErstelltAmDesc(lieferantId)
        return ResponseEntity.ok(list.map(::mapToDto))
    }

    @PatchMapping("/{id}/status")
    @Transactional
    fun updateStatus(@PathVariable id: Long, @RequestParam status: ReklamationStatus): ResponseEntity<Void> {
        val reklamation = reklamationRepository.findById(id).orElse(null)
            ?: return ResponseEntity.notFound().build()
        reklamation.status = status
        reklamationRepository.save(reklamation)
        return ResponseEntity.ok().build()
    }

    @DeleteMapping("/{id}")
    @Transactional
    fun delete(@PathVariable id: Long): ResponseEntity<Void> {
        if (!reklamationRepository.existsById(id)) {
            return ResponseEntity.notFound().build()
        }
        reklamationRepository.deleteById(id)
        return ResponseEntity.noContent().build()
    }

    @PostMapping("/lieferant/{lieferantId}")
    @Transactional
    fun create(
        @PathVariable lieferantId: Long,
        @RequestBody request: CreateReklamationRequest,
        @RequestParam(required = false) token: String?,
    ): ResponseEntity<LieferantReklamationDto> {
        val lieferant = lieferantenRepository.findById(lieferantId).orElse(null)
            ?: return ResponseEntity.notFound().build()
        val mitarbeiter = mitarbeiterByToken(token)

        var reklamation = LieferantReklamation().apply {
            this.lieferant = lieferant
            erstelltVon = mitarbeiter
            beschreibung = request.beschreibung
            status = request.status ?: ReklamationStatus.OFFEN
        }

        request.lieferscheinId?.let { lieferscheinId ->
            val lieferschein = dokumentRepository.findById(lieferscheinId).orElse(null)
            if (lieferschein != null && lieferschein.lieferant?.id == lieferantId) {
                reklamation.lieferschein = lieferschein
            }
        }

        reklamation = reklamationRepository.save(reklamation)
        return ResponseEntity.status(HttpStatus.CREATED).body(mapToDto(reklamation))
    }

    @PostMapping(value = ["/{id}/bilder"], consumes = [MediaType.MULTIPART_FORM_DATA_VALUE])
    @Transactional
    fun uploadBild(
        @PathVariable id: Long,
        @RequestPart("datei") datei: MultipartFile,
        @RequestParam(required = false) token: String?,
    ): ResponseEntity<LieferantBildDto> {
        val reklamation = reklamationRepository.findById(id).orElse(null)
            ?: return ResponseEntity.notFound().build()
        val mitarbeiter = mitarbeiterByToken(token)

        return try {
            val originalFilename = StringUtils.cleanPath(Objects.requireNonNull(datei.originalFilename))
            val storedFilename = "${UUID.randomUUID()}_$originalFilename"
            val lieferantDir = Paths.get("uploads", "lieferanten", reklamation.lieferant?.id.toString(), "bilder")
            Files.createDirectories(lieferantDir)
            val targetPath = lieferantDir.resolve(storedFilename)
            datei.transferTo(targetPath)

            var bild = LieferantBild().apply {
                lieferant = reklamation.lieferant
                this.reklamation = reklamation
                originalDateiname = originalFilename
                gespeicherterDateiname = storedFilename
                erstelltAm = LocalDateTime.now()
                hochgeladenVon = mitarbeiter
            }
            bild = bildRepository.save(bild)

            ResponseEntity.status(HttpStatus.CREATED).body(toBildDto(bild, mitarbeiter))
        } catch (_: IOException) {
            ResponseEntity.internalServerError().build()
        }
    }

    @GetMapping("/lieferscheine/search")
    fun searchLieferscheine(
        @RequestParam lieferantId: Long,
        @RequestParam query: String,
    ): ResponseEntity<List<LieferscheinSearchDto>> {
        val results = dokumentRepository.searchLieferscheine(lieferantId, query)
        val dtos = results.map { doc ->
            LieferscheinSearchDto().apply {
                id = doc.id
                originalDateiname = doc.getEffektiverDateiname()
                if (doc.geschaeftsdaten != null) {
                    dokumentNummer = doc.geschaeftsdaten?.dokumentNummer
                    datum = doc.geschaeftsdaten?.dokumentDatum
                } else {
                    datum = doc.uploadDatum?.toLocalDate()
                }
            }
        }
        return ResponseEntity.ok(dtos)
    }

    private fun mapToDto(entity: LieferantReklamation): LieferantReklamationDto =
        LieferantReklamationDto().apply {
            id = entity.id
            lieferantId = entity.lieferant?.id
            lieferantName = entity.lieferant?.lieferantenname

            entity.lieferschein?.let { lieferschein ->
                lieferscheinId = lieferschein.id
                lieferscheinDateiname = lieferschein.getEffektiverDateiname()
                lieferscheinNummer = lieferschein.geschaeftsdaten?.dokumentNummer
            }

            erstellerName = entity.erstelltVon?.let { "${it.vorname} ${it.nachname}" }
            erstelltAm = entity.erstelltAm
            beschreibung = entity.beschreibung
            status = entity.status
            bilder = entity.bilder.map { toBildDto(it, it.hochgeladenVon) }
        }

    private fun toBildDto(bild: LieferantBild, mitarbeiter: Mitarbeiter?): LieferantBildDto =
        LieferantBildDto().apply {
            id = bild.id
            originalDateiname = bild.originalDateiname
            url = "/api/lieferanten/bilder/file/${bild.gespeicherterDateiname}"
            beschreibung = bild.beschreibung
            erstelltAm = bild.erstelltAm
            if (mitarbeiter != null) {
                mitarbeiterVorname = mitarbeiter.vorname
                mitarbeiterNachname = mitarbeiter.nachname
            }
        }

    private fun mitarbeiterByToken(token: String?): Mitarbeiter? {
        if (!StringUtils.hasText(token)) {
            return null
        }
        return mitarbeiterRepository.findByLoginToken(token!!).orElse(null)
    }

    class LieferscheinSearchDto {
        var id: Long? = null
        var dokumentNummer: String? = null
        var originalDateiname: String? = null
        var datum: LocalDate? = null
    }
}
